share/dkg/dkg.go

```diff
@@ -0,0 +1,660 @@
+//    `SecretCommits()`. If there is an error, it can return a commitment complaint
+//    (ComplaintCommits) that must be broadcasted to the QUAL set.
+//   6. Each QUAL participant receiving a complaint can process it with
+//    `ProcessComplaintCommits()` which returns the secret share
+//    (ReconstructCommits) given from the malicious participant. This structure
+//    must be broadcasted to all the QUAL participant.
+//   7. At this point, every QUAL participant can issue the distributed key by
+//    calling `DistKeyShare()` !
+package dkg
+
+import (
+	"bytes"
+	"crypto/cipher"
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/share"
+	"github.com/dedis/crypto/share/vss"
+	"github.com/dedis/crypto/sign"
+)
+
+// DistKeyShare holds the share of a distributed key for a participant.
+type DistKeyShare struct {
+	// Coefficients of the public polynomial holding the public key
+	Commits []abstract.Point
+	// share of the distributed secret
+	Share *share.PriShare
+}
+
+// Public returns the public key associated with the distributed private key.
+func (d *DistKeyShare) Public() abstract.Point {
+	return d.Commits[0]
+}
+
+// Deal holds the Deal for one participants as well as the index of the issuing
+// Dealer.
+//  NOTE: Doing that in vss.go would be possible but then the Dealer is always
+//  assumed to be a member of the participants. It's only the case here.
+type Deal struct {
+	// Index of the Dealer in the list of participants
+	Index uint32
+	// Deal issued for another participant.
+	Deal *vss.EncryptedDeal
+}
+
+// Response holds the Response from another participants as well as the index of
+// the targeted Dealer.
+type Response struct {
+	// Index of the Dealer for which this response is for
+	Index uint32
+	// Response issued from another participant.
+	Response *vss.Response
+}
+
+// Justification holds the Justification from a Dealer as well as the index of
+// the Dealer in question.
+type Justification struct {
+	// Index of the Dealer who answered with this Justification
+	Index uint32
+	// Justification issued from the Dealer.
+	Justification *vss.Justification
+}
+
+// SecretCommits is sent during the distributed public key reconstruction phase,
+// basically a Feldman VSS scheme.
+type SecretCommits struct {
+	// Index of the Dealer in the list of participants
+	Index uint32
+	// Commitments generated by the Dealer
+	Commitments []abstract.Point
+	// SessionID generated by the Dealer tied to the Deal.
+	SessionID []byte
+	// Signature from the Dealer
+	Signature []byte
+}
+
+// ComplaintCommits is sent if the secret commitments revealed by a peer are not
+// valid.
+type ComplaintCommits struct {
+	// Index of the Verifier _issuing_ the ComplaintCommit
+	Index uint32
+	// DealerIndex being the index of the Dealer who issued the SecretCommits
+	DealerIndex uint32
+	// Deal that has been given from the Dealer (at DealerIndex) to this node
+	// (at Index)
```

Comment:
In RandHound I think I called those indices `source` and `target` or `src` and `tgt` to better differentiate between them. I am also fine with the current notion above. Just as food for thought.